Windows: Add Direct3D texture interoperability support

jnschulze opened the pull request:
This PR adds support for Direct3D texture interop using ANGLE's `ANGLE_d3d_share_handle_client_buffer` extension (see https://www.khronos.org/registry/EGL/extensions/ANGLE/EGL_ANGLE_d3d_share_handle_client_buffer.txt)

It adds a new texture variant, `flutter::GpuSurfaceTexture`, as an addition to the existing `flutter::PixelBufferTexture`.

See https://github.com/Microsoft/angle/wiki/Interop-with-other-DirectX-code for an example of how to create compatible Direct3D textures.

## Pre-launch Checklist

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [x] I listed at least one issue that this PR fixes in the description above.
- [x] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] I signed the [CLA].
- [x] All existing and new tests are passing.
- [x] The reviewer has submitted any presubmit flakes in this PR using the [engine presubmit flakes form] before re-triggering the failure.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

<!-- Links -->
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/master/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat
[engine presubmit flakes form]: https://forms.gle/Wc1VyFRYJjQTH6w5A

chinmaygarde commented:
cc @cbracken @clarkezone. The approach looks fine to me. But the presubmit failures are real though ([compilation failures](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8843892564173401104/+/u/Build_and_test_arm64_profile/build_host_profile_unopt/stdout)). Can you please address those in the meantime?

jnschulze commented:
@chinmaygarde
> But the presubmit failures are real though ([compilation failures](https://logs.chromium.org/logs/flutter/buildbucket/cr-buildbucket.appspot.com/8843892564173401104/+/u/Build_and_test_arm64_profile/build_host_profile_unopt/stdout)). Can you please address those in the meantime?

done.

clarkezone commented:
I'll take a look over the weekend

clarkezone commented:
Took a quick look, looks good based on that.

chinmaygarde commented:
cc @clarkezone Care to approve the patch 😄

clarkezone commented:
@jnschulze do you have a test app I can run against your branch?

jnschulze commented:
@clarkezone
Yes, this webview_windows should work: https://github.com/jnschulze/flutter-webview-windows/commits/features/gpu-surface-texture

chinmaygarde commented:
@clarkezone Do you have the cycles to look over the sample?

clarkezone commented:
Appols.. I have not had a chance to look and I'm heading out on a vacation for a week.  I have cycles on 7/26 that I could carve out to look but appreciate that is a ways out there.

knopp commented on shell/platform/common/public/flutter_texture_registrar.h:
```diff
@@ -36,6 +58,26 @@ typedef struct {
+  // The surface handle.
+  // For DirectX textures (kFlutterDesktopGpuSurfaceTypeDxgi), this is the
+  // shared handle of an IDXGIResource.
+  void* handle;
```
What is the ownership policy here? Does engine take over the reference, or does it increases the reference count?

knopp replied:
I guess the surface descriptor being passed as const sort of answers my questions.

jnschulze replied:
ANGLE increments the reference count once the shared handle has been opened.

However, as this happens asynchronously due to the GPU surface descriptor callback, it's advised to explicitly increment the texture's reference count before the callback returns in order to prevent potential races (a plugin might destroy the texture after the callback returns but just before the Flutter texture gets populated). This obviously requires a `FlutterDesktopGpuSurfaceDescriptor::release_callback` keeping the reference count in balance.

knopp commented:
@jnschulze, I tried the gpu-surface-texture webview branch, and I have a questions. The fallback texture bridge seems to work fine, but when enabling the gpu surface (`use_gpu_surface=true` in `WebviewBridge` constructor, the surface seem to have scaling issues (it doesn't fill the entire window). How much of window it fills seems to depend on window size. Are you aware of this? 

<img width="844" alt="Screen Shot 2021-07-18 at 2 39 38 PM" src="https://user-images.githubusercontent.com/96958/126067426-d5d31ef9-683c-4596-bcbb-7110d858ea2d.png">

jnschulze commented:
@knopp 
Thanks, I will take a look this week.

chinmaygarde commented:
@jnschulze Any updates?

jnschulze commented:
@chinmaygarde
Sorry, I haven't had a chance to see what's wrong with the example yet.

clarkezone commented:
@jnschulze finally had a chance to test this out.  Worked with correct scaling.  I don't have an opinion on your question above, I'd be good either way.

clarkezone reviewed (approved):
LGTM

ds84182 commented:
Any updates on this PR? It's already been given an LGTM, but hasn't been merged yet. Being able to import textures from D3D opens up a lot of opportunities for plugins on Windows.

oddko commented:
Very interested by this for real time 3D rendering applications, pixel buffers currently do not cut it and Windows so far is the only platform where GPU-backed buffers are not available.

jnschulze commented:
I’ve been using the proposed interface for more than 6 months in production with both `webview_windows` and `libvlc`. However, this PR is a bit outdated as I did one or two tweaks here and there. I will push an update this week.

ZhenFTW commented:
What's taking too long for this PR to get merged? Is it not priority ? Having this merged will provide more flutter windows plugin including smoother video playback.

jnschulze commented:
I've just pushed an update.

#### Changes since the previous commit

- `kFlutterDesktopGpuSurfaceTypeDxgi` was renamed to `kFlutterDesktopGpuSurfaceTypeDxgiSharedHandle`
- `kFlutterDesktopGpuSurfaceTypeD3d11Texture2D` was added
- `release_callback` and `release_context` were added to  `FlutterDesktopGpuSurfaceDescriptor`
- `struct_size` was added to  `FlutterDesktopGpuSurfaceDescriptor` in order to safely support future changes.
- Tests were added

jnschulze commented:
> Why not exposing OpenGL and instead going straight for D3D interop though ?

Because Flutter effectively uses DirectX on Windows. The fact that it uses ANGLE for translating OpenGL to D3D on Windows is an implementation detail which might change in the future.
Besides, exposing OpenGL would require additional APIs for context sharing etc.

oddko commented:
> Because Flutter effectively uses DirectX on Windows. The fact that it uses ANGLE for translating OpenGL to D3D on Windows is an implementation detail which might change in the future.

True but so is the use of DirectX on Windows which is itself an implementation detail of ANGLE on Windows. My comment was more on the fact that you have (as it seems from the PR) to go through ANGLE anyway for the interop, so OpenGLES seemed like an easier target. But given you used this to embed native views (like webviews) that makes sense you used D3D.

alexmercerind commented:
Greetings!

I need a little help with this. So, I'm trying to render stuff using open-gl APIs with this texture. By using ANGLE & using shared `HANDLE` after creating `ID3D11Device` & `ID3D11Texture2D`. But, everything is just blank. Am I doing something completely meaningless? Is this not how it's meant to be used?

I just stitched together various code I could find related to ANGLE, Direct3D or this new Direct3D texture API in Flutter. There's no boilerplate, everything is written very procedurally to have a minimal running example. 

I've configured everything in the repository & a single `flutter run` will work. Your help/guidance will be appreciated.

Thankyou.

Repository: https://github.com/alexmercerind/flutter_windows_angle_d3d_texture
Dart: https://github.com/alexmercerind/flutter_windows_angle_d3d_texture/blob/5651c198b0ab6aeb7475d5da820d7bd1bebbf44d/example/lib/main.dart#L25-L35
C++ Plugin Interface: https://github.com/alexmercerind/flutter_windows_angle_d3d_texture/blob/5651c198b0ab6aeb7475d5da820d7bd1bebbf44d/windows/flutter_windows_angle_d3d_texture_plugin.cpp#L102-L314

![Screenshot 2022-07-07 202945](https://user-images.githubusercontent.com/28951144/177807420-e52e3c4f-f6b4-44e2-82bb-4b9973e93444.png)


Used References

https://github.com/Microsoft/angle/wiki/Interop-with-other-DirectX-code
https://github.com/google/angle/blob/main/samples/hello_triangle/HelloTriangle.cpp
https://github.com/clarkezone/anglehosting
